### DISCUSSION TITLE
Why does the C++ STL mix tabs and spaces?

### DISCUSSION THREAD

**Discussion by BrandonPacewic** [microsoft/STL#2725] [Q&A] [answered]
When looking at the source code for STL containers like `vector` I realized that most of the formatting makes no sense to me. It appears that the indent width of class functions is 6 spaces. Then once inside the function it switches to 1 tab? Then for anything nested inside that function it adds 2 spaces?

This causes some problems in my editor where it looks like the first items nested in a class function
have no indentation at all. So why is this done?

**Comment by miscco** [comment, marked as answer]
That looks like libstdc++ code.

This is MSVC stl and the code here looks much different

**Reply by StephanTLavavej** [reply]
Yep. In fact, we have automated validation that rejects tabs in virtually all files (see https://github.com/microsoft/STL/blob/313aa593223e78b458ac50a5a46357601eebfe00/tools/validate/validate.cpp#L227-L229 for the current lone exception).

We converted from tabs to spaces in VS 2019 16.0, when we switched to clang-formatting everything.

**Comment by BrandonPacewic** [comment]
Looking again I realized that this is only true for the GNU ISO C++ Library. I suppose that my question would then be why does GNU do this? But yes the MSVC STL does not do this. My apologies for not realizing that I was looking at the wrong library.

**Reply by miscco** [reply]
This forum is only about MSVC stl. Moreover, all maintainers are explicitly forbidden to even look at libstdc++ source code due to lisencing issues. 

So it would be nice if you would both remove the snippet in question and pose the question in their mailing list
